### DISCUSSION TITLE
Linking webots to python environment

### DISCUSSION THREAD

**Discussion by perceptualrobots** [cyberbotics/wrestling#32] [Q&A] [answered]
I would like to run the simulation loop without the gui. I have an anaconda environment where I have my own libraries, and others installed with pip. Can I install the webots package here?

I would also like to run things in the gui. I have installed R2023a. How do I connect it to my anaconda environment? 

In preferencs I have set "Python command" to C:\Users\ryoung\AppData\Local\Programs\Python\Python39\python.exe. This lets me run the loop from the gui, but doesn't recognise other pacakges. How do I install packages that can be found from pip or on my local machine?

What am I missing?

Rupert

**Comment by omichel** [comment, marked as answer]
Hi Rupert,

First, I would recommend you to install a recent [nightly build of Webots R2023a](https://github.com/cyberbotics/webots/releases/), as R2023a has some bugs with Python forcing you to restart Webots if a Python error occurred in your controller program.

Normally, from your Python environment, you should be able to pip install Python modules. If configured properly, these modules should found when starting your controllers from Webots. If not, please double-check that the `PYTHONPATH` global environment variable points actually to the place where your Python modules are installed.

I don't have any experience with anaconda Python, but I never had such problems with the official Python distributions (from https://python.org). Would it be possible possible for you to use this official version of Python?

**Reply by perceptualrobots** [reply]
Hi Olivier,

> First, I would recommend you to install a recent [nightly build of Webots R2023a](https://github.com/cyberbotics/webots/releases/), 

I saw those, though couldn't spot the install file.

> I don't have any experience with anaconda Python, but I never had such problems with the official Python distributions (from https://python.org). Would it be possible possible for you to use this official version of Python?

I have set up a separate environment on Windows, by installing python and adding it to the PATH variable. I also installed pip. From Windows command prompt I installed standard libraries with pip.

> Normally, from your Python environment, you should be able to pip install Python modules. If configured properly, these modules should found when starting your controllers from Webots. If not, please double-check that the `PYTHONPATH` global environment variable points actually to the place where your Python modules are installed.

I tried to install my own two libraries by changing to their directories and running "pip install -e .". For some reason only one of these was recognised within webots. So, I added both paths to PYTHONPATH and that seemed to do the trick.

So, now I can acces my own modules from the webots GUI, though how about my other query, how do I access the webots classes/functions (e.g. controller, MotionLibrary) from outside the GUI? I tried to install from the cloned wrestling (called icra2023) directory with  "pip install -e .", but get this error "ERROR: file:///C:/Users/ryoung/Versioning/icra2023 does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found."

**Reply by omichel** [reply]
> I saw those, though couldn't spot the install file.

The install file is located in the assets section (which you can open by clicking on it). Here is a direct link to the one of Feb 17, 2023: [webots-R2023a-rev1_setup.exe](https://github.com/cyberbotics/webots/releases/download/nightly_17_2_2023/webots-R2023a-rev1_setup.exe) (note: it will expire in 2 or 3 days).

> So, now I can acces my own modules from the webots GUI, though how about my other query, how do I access the webots classes/functions (e.g. controller, MotionLibrary) from outside the GUI?

Again, it is a matter to setting the `PYTHONPATH` and `PATH` environment variables so that Python will be able find and load these modules. See the [Webots documentation on how to set-up an extern controller in Webots](https://cyberbotics.com/doc/guide/running-extern-robot-controllers#environment-variables). This is actually useful only if you plan to run your controller as an extern controller. Otherwise, when Webots starts your Python controller, it will do all this automatically for you.

**Reply by perceptualrobots** [reply]
> The install file is located in the assets section (which you can open by clicking on it). Here is a direct link to the one of Feb 17, 2023: [webots-R2023a-rev1_setup.exe](https://github.com/cyberbotics/webots/releases/download/nightly_17_2_2023/webots-R2023a-rev1_setup.exe) (note: it will expire in 2 or 3 days).

Thanks, got that.

> Again, it is a matter to setting the `PYTHONPATH` and `PATH` environment variables so that Python will be able find and load these modules. See the [Webots documentation on how to set-up an extern controller in Webots](https://cyberbotics.com/doc/guide/running-extern-robot-controllers#environment-variables). This is actually useful only if you plan to run your controller as an extern controller. Otherwise, when Webots starts your Python controller, it will do all this automatically for you.

Ok did that. Though I think the webots_home variable for PYTHONPATH for Windows was showing the unix version, so I changed ${WEBOTS_HOME} to %WEBOTS_HOME%.

Now my participant.py recognises "from controller import Robot". Though it is not recognising from "utils.motion_library import MotionLibrary".

`ModuleNotFoundError: No module named 'utils'`

Where would I find that?

The reason I am doing this is because I want to run hundreds of loops  for a learning algorithm and I am assuming that each loop would 
 run much faster outside of the GUI. Is that assumption correct?

**Reply by omichel** [reply]
The `utils` folder is located in your repository. See for example the [utils](https://github.com/cyberbotics/wrestling-fatima/tree/main/controllers/utils) folder of the sample controller. You should have a similar folder in your local installation. Simply adding the path to this folder in your `PYTHONPATH` environment variable should fix the problem.

**Reply by perceptualrobots** [reply]
> The `utils` folder is located in your repository. See for example the [utils](https://github.com/cyberbotics/wrestling-fatima/tree/main/controllers/utils) folder of the sample controller. You should have a similar folder in your local installation. Simply adding the path to this folder in your `PYTHONPATH` environment variable should fix the problem.

Ah, ok, thanks, didn't see that. 

Now getting this message:
`Cannot connect to Webots instance, retrying for another 50 seconds...`

What am I missing?

**Reply by omichel** [reply]
I order to connect your controller as an extern controller, you should also edit the world file and change the [controller of your robot](https://github.com/cyberbotics/wrestling/blob/main/worlds/wrestling.wbt#L118) from `"participant"` to `"<extern>"`.

**Reply by perceptualrobots** [reply]
> I order to connect your controller as an extern controller, you should also edit the world file and change the [controller of your robot](https://github.com/cyberbotics/wrestling/blob/main/worlds/wrestling.wbt#L118) from `"participant"` to `"<extern>"`.

Thanks, I am now getting the 'connected' message in the webots referee console. I am running a basci wrestler, but nothing is happening. It seems to be stuck on the while line. I do not get the print output.

Am I missing something?

```
 while self.step(time_step) != -1:  # mandatory function to make the simulation run
            print("b4")
            motion_library.play('Forwards')
            print("b5")
```

**Reply by perceptualrobots** [reply]
> Thanks, I am now getting the 'connected' message in the webots referee console. I am running a basci wrestler, but nothing is happening. It seems to be stuck on the while line. I do not get the print output.
> 
> Am I missing something?
> 
> ```
>  while self.step(time_step) != -1:  # mandatory function to make the simulation run
>             print("b4")
>             motion_library.play('Forwards')
>             print("b5")
> ```

This is working now. I am trying to set up a couple of different environments, so I amy have been connecting from the wrong one or something.
